src/rt/arc.rs

```diff
@@ -117,16 +117,24 @@ impl Arc {
             }
         })
     }
+
+    fn branch(&self, action: Action) {
+        let r = self.state;
+        r.branch_action(action);
+        assert!(
+            r.ref_eq(self.state),
+            "Internal state mutated during branch. This is \
+                usually due to a bug in the algorithm being tested writing in \
+                an invalid memory location."
```

Comment:
Do you think this should include the caller that allocated the `Arc`?

Reply:
I'm not trying to provide exhaustive location tracking in this PR. Follow up PRs can do more work on that front.

Reply:
cool, fair enough!